ethereum/contracts/BasicInboundChannelV2.sol

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.5;
+pragma experimental ABIEncoderV2;
+
+import "./ParachainLightClient.sol";
+import "./BeefyLightClient.sol";
+import "./SimplifiedMMRVerification.sol";
+
+contract BasicInboundChannelV2 {
+    uint256 public constant MAX_GAS_PER_MESSAGE = 100000;
+    uint256 public constant GAS_BUFFER = 60000;
+
+    mapping(bytes32 => uint64) public userNonce;
+
+    BeefyLightClient public beefyLightClient;
+
+    struct Leaf {
+        bytes32 account;
+        uint64 nonce;
+        Message[] messages;
+    }
+
+    struct Message {
+        address target;
+        bytes payload;
+    }
+
+    event MessageDispatched(uint64 nonce, bool result);
+
+    constructor(BeefyLightClient _beefyLightClient) {
+        beefyLightClient = _beefyLightClient;
+    }
+
+    function generateCommitmentHash(Leaf calldata _leaf,bytes32[] calldata _leafProof)
```

Comment:
we already have and use a MerkleTree library. Would we be able to use that instead of re-implementing this?

Reply:
updated